### PR TITLE
Checkout with dry run

### PR DESCRIPTION
**Pull request by bpeabody**
Should address https://github.com/twosigma/git-meta/issues/104

### CODE REVIEW

File: node/lib/util/checkout.js [hunk 1]
```diff
@@ -62,16 +64,45 @@ exports.checkout = co.wrap(function *(metaRepo, committish) {
...
+    }
+
+    // Try the submodules; store the opened repos and loaded commits for use
+    // in the actual checkout later.
+
+    let subRepos = [];     // will contain a list of repositories for sub-repos
+    let subCommits = [];   // will contain a list of commits to checkout
 
     yield open.map(co.wrap(function *(name) {
         // Open repo but not alive on this commit.
```
Comment (novalis) [inline node/lib/util/checkout.js]: How do we actually want to handle this case?  This is, in effect, checking out a deletion, right?

File: node/lib/util/checkout.js [hunk 2]
```diff
@@ -81,10 +112,51 @@ exports.checkout = co.wrap(function *(metaRepo, committish) {
...
+
+    if (0 !== errors.length) {
+        throw new UserError(errors.join("\n"));
+    }
+
+    /**
+     * Checkout and set as head the specified `commit` in the specified `repo`.
+     */
+    const doCheckout = co.wrap(function *(repo, commit) {
+        yield NodeGit.Checkout.tree(repo, commit, {
```
Comment (novalis) [inline node/lib/util/checkout.js]: I'm confused by running checkout twice.

File: node/lib/util/checkout.js [hunk 3]
```diff
@@ -81,10 +112,51 @@ exports.checkout = co.wrap(function *(metaRepo, committish) {
...
+        });
+        repo.setHeadDetached(commit);
+        yield NodeGit.Checkout.head(repo, {
+            checkoutStrategy: NodeGit.Checkout.STRATEGY.SAFE,
+        });
+    });
+
+    // Now do the actual checkouts.
+
+    yield doCheckout(metaRepo, commit);
```
Comment (novalis) [inline node/lib/util/checkout.js]: A safe checkout strategy on the meta repo is going to refuse to touch modified submodules, and thus is going to fail, right?

File: node/lib/util/checkout.js (hunk 1 above)
Comment (bpeabody) [inline node/lib/util/checkout.js]: Yes, I guess we should close the repo or else it's going to look like a change to Git. Does that sound reasonable?

File: node/lib/util/checkout.js (hunk 3 above)
Comment (bpeabody) [inline node/lib/util/checkout.js]: It appears to not have that effect. I'll add a test case to verify.

File: node/lib/util/checkout.js (hunk 2 above)
Comment (bpeabody) [inline node/lib/util/checkout.js]: The only way I could make it behave like `git checkout` was to run it twice; basically, once to apply to the index and once to the workdir. I'll add a comment to explain.

File: node/lib/util/checkout.js (hunk 1 above)
Comment (bpeabody) [inline node/lib/util/checkout.js]: Which makes me wonder if we need to do the same thing after reset, rebase, etc. Let's leave this as is for now and discuss. Whatever we decide to do I'll apply uniformly to related functions.

File: node/lib/util/checkout.js (hunk 3 above)
Comment (bpeabody) [inline node/lib/util/checkout.js]: OK, I added some test cases to verify -- `checkout` does not view any *committed* changes in submodes as being unsafe.

**Comment by bpeabody** [comment]
Think we're good now.

File: node/lib/util/checkout.js (hunk 3 above)
Comment (novalis) [inline node/lib/util/checkout.js]: That is, nonconflicting committed changes, I assume.

Comment (bpeabody) [inline node/lib/util/checkout.js]: The SAFE option appears to have no effect on the state of the submodules whatsoever.

**Pull request merged**